### ISSUE TITLE
发现一个小bug，但是比较严重的问题，建议作者修改一下prompt

### ISSUE THREAD

**Issue by egrcc** [LianjiaTech/BELLE#337]
我使用的是老版代码，
```
input_text = "Human: " + instruction + input_text + "\n\nAssistant: "
```
这里最后一位的空格，如果是只将input_text来tokenize的话，会成为一个单独的token，
但是拼上target_text后，如果将full_prompt来tokenize，空格有可能会跟后面一位合并成一个token，导致在mask的时候出现多mask了一位。
<img width="872" alt="截屏2023-05-06 11 09 34" src="https://user-images.githubusercontent.com/6469953/236595636-1daa565f-f6e0-4dfd-813c-9f896ef0595e.png">
下面是一个例子：
```
data_point = {"instruction": "请判断是否正确", "input": "", "output": "是"}
```
此时，如果只将input_text来tokenize，会得到：

{'input_ids': [1, 1, 12968, 29901, 37033, 34301, 32520, 33325, 13, 13, 7900, 22137, 29901, **29871**], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], 'labels': [1, 1, 12968, 29901, 37033, 34301, 32520, 33325, 13, 13, 7900, 22137, 29901, 29871]}

如果将full_prompt来tokenize，会得到：

{'input_ids': [1, 1, 12968, 29901, 37033, 34301, 32520, 33325, 13, 13, 7900, 22137, 29901, **34043**, 2], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], 'labels': [1, 1, 12968, 29901, 37033, 34301, 32520, 33325, 13, 13, 7900, 22137, 29901, 34043, 2]}

<img width="371" alt="截屏2023-05-06 11 35 11" src="https://user-images.githubusercontent.com/6469953/236596837-8a0b973a-755e-47de-af48-08b3ab1edf1f.png">

这就会导致，如果下游任务是回答”是否“这种类型，在微调的时候loss会变成0，因为”34043“这个token也被mask了。

如果作者的新版代码还是使用的同样的prompt的话，建议修改一下。

**Comment by xianghuisun** [comment]
> 我使用的是老版代码，
> 
> ```
> input_text = "Human: " + instruction + input_text + "\n\nAssistant: "
> ```
> 
> 这里最后一位的空格，如果是只将input_text来tokenize的话，会成为一个单独的token， 但是拼上target_text后，如果将full_prompt来tokenize，空格有可能会跟后面一位合并成一个token，导致在mask的时候出现多mask了一位。 <img alt="截屏2023-05-06 11 09 34" width="872" src="https://user-images.githubusercontent.com/6469953/236595636-1daa565f-f6e0-4dfd-813c-9f896ef0595e.png"> 下面是一个例子：
> 
> ```
> data_point = {"instruction": "请判断是否正确", "input": "", "output": "是"}
> ```
> 
> 此时，如果只将input_text来tokenize，会得到：
> 
> {'input_ids': [1, 1, 12968, 29901, 37033, 34301, 32520, 33325, 13, 13, 7900, 22137, 29901, **29871**], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], 'labels': [1, 1, 12968, 29901, 37033, 34301, 32520, 33325, 13, 13, 7900, 22137, 29901, 29871]}
> 
> 如果将full_prompt来tokenize，会得到：
> 
> {'input_ids': [1, 1, 12968, 29901, 37033, 34301, 32520, 33325, 13, 13, 7900, 22137, 29901, **34043**, 2], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], 'labels': [1, 1, 12968, 29901, 37033, 34301, 32520, 33325, 13, 13, 7900, 22137, 29901, 34043, 2]}
> 
> <img alt="截屏2023-05-06 11 35 11" width="371" src="https://user-images.githubusercontent.com/6469953/236596837-8a0b973a-755e-47de-af48-08b3ab1edf1f.png">
> 
> 这就会导致，如果下游任务是回答”是否“这种类型，在微调的时候loss会变成0，因为”34043“这个token也被mask了。
> 
> 如果作者的新版代码还是使用的同样的prompt的话，建议修改一下。

已修改。非常感谢您提出的问题。

**Comment by ZeyuTeng96** [comment]
思考了一个问题，如果只是把空格给删去，还是无法保证Assistant:这个结尾的”:“不和target_text里的词合并成一个token
@egrcc @xianghuisun

**Comment by xianghuisun** [comment]
> 思考了一个问题，如果只是把空格给删去，还是无法保证Assistant:这个结尾的”:“不和target_text里的词合并成一个token @egrcc @xianghuisun

sentence_value = 'Human: \n' + sentence["value"] + '\n\nAssistant: \n' if sentence_from == 'human' else sentence["value"] #https://github.com/LianjiaTech/BELLE/issues/337

新版代码的prompt

**Comment by xianghuisun** [comment]
> 思考了一个问题，如果只是把空格给删去，还是无法保证Assistant:这个结尾的”:“不和target_text里的词合并成一个token @egrcc @xianghuisun

特意加了换行符

**Comment by lucasjinreal** [comment]
@xianghuisun 想请教一下，看蛮多实现是没有 换行的，例如Billa：

```
for example in examples:
            prefix = "Human: "
            for key in example:
                example[key] = example[key].strip()
            assert len(example["instruction"]) + len(example["input"]) > 0 
            if len(example["instruction"]) > 0:
                prefix += example["instruction"] + "\n"
            if len(example["input"]) > 0:
                prefix += example["input"] + "\n"
            prefix += "Assistant: "
```

这两种格式到底谁更优一点？

**Comment by Luoyingfeng8** [comment]
> 
大佬，没太理解什么意思，sentencepiece不是把空格当做一个字符吗，这块跟mask是什么关系

**Comment by egrcc** [comment]
> > 
> 
> 大佬，没太理解什么意思，sentencepiece不是把空格当做一个字符吗，这块跟mask是什么关系

@Luoyingfeng8 这个空格有时会跟后面的字合并成一个token，导致mask的时候会多mask一位
